src/solver/nllloss/forward_nllloss.cpp

```diff
@@ -0,0 +1,138 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include "miopen/conv_solution.hpp"
+#include "miopen/execution_context.hpp"
+#include "miopen/invoke_params.hpp"
+#include "miopen/kernel_info.hpp"
+#include <cstddef>
+#include <miopen/nllloss/solvers.hpp>
+
+#include <miopen/nllloss/invoke_params.hpp>
+#include <miopen/datatype.hpp>
+#include <miopen/nllloss.hpp>
+#include <miopen/kernel_build_params.hpp>
+#include <miopen/target_properties.hpp>
+
+#define LOCAL_SIZE 1024
+
+namespace miopen {
+
+namespace solver {
+
+namespace nllloss {
+
+bool NLLLossForward::IsApplicable(const ExecutionContext&,
+                                  const miopen::nllloss::ProblemDescription& problem) const
+{
+    if(!problem.IsSameType())
+        return false;
+    if(!problem.IsAllPacked())
+        return false;
+    if(!problem.IsRightDim())
+        return false;
+    return true;
+}
+
+ConvSolution NLLLossForward::GetSolution(const ExecutionContext& context,
+                                         const miopen::nllloss::ProblemDescription& problem) const
+{
+    std::ignore = context;
+
+    auto result       = ConvSolution{miopenStatusSuccess};
+    auto input_dtype  = miopen::GetDataType(problem.GetInputDesc().GetType());
+    auto output_dtype = miopen::GetDataType(problem.GetOutputDesc().GetType());
+
+    {
+        auto dtype   = problem.GetInputDesc().GetType();
+        long N_total = problem.GetNtotal();
+
+        size_t xlocalsize = LOCAL_SIZE;
+        size_t xgridsize  = N_total + LOCAL_SIZE - 1;
```

Comment:
maybe.. align up?

```
in C++
(N_total + LOCAL_SIZE - 1) != (N_total + LOCAL_SIZE - 1) / LOCAL_SIZE  * LOCAL_SIZE
```
example
1000 + 256 - 1 = 1255
(1000 + 256 - 1) / 256 * 256 = 1024

Reply:
Yes, I haven't worked on it carefully. I changed it to (N_total + LOCAL_SIZE - 1) / LOCAL_SIZE  * LOCAL_SIZE. Is it ok or we have a defined align up function?

Reply:
Oh I saw the function, sorry for the confusing